doc/astroid.1.scd
astroid(1)

# NAME

astroid - graphical threads-with-tags style, lightweight and fast, e-mail
client for Notmuch

# SYNOPSIS

*astroid* [options]

# DESCRIPTION

Astroid is a lightweight and fast *Mail User Agent* that provides a graphical
interface to searching, displaying and composing email, organized in threads
and tags. It uses the Notmuch backend for blazingly fast searches through
tons of e-mail.

# OPTIONS

*-c, --config* <config>
	Override the default configuration path. Please refer to the *CONFIGURATION*
	section for details.

*--disable-log*
	Disable logging.

*-h, --help*
	Show the help message and quit.

*--log-level <level>*
	Override the configured log level. Level options are: trace, debug (default),
	info, warn, error and fatal.

*--log-stdout*
	Output logs to the standard output, regardless of any other logging
	configuration.

*-m, --mailto* <dest>
	Open the composition window, sending to the <dest> url or address.

*-n, --new-config*
	Generate a new configuration file and quit.

*--no-auto-poll*
	Disable automatic polling.

*--refresh* <revision>
	Update the user view of a running astroid instance with any changes detected
	in the mail directory since <revision>. You can obtain the current revision
	with *notmuch count --lastmod | cut -f3* (*0* will refresh all
	thread-indexes). --{start,stop}-polling can be used as an *alternative*, but
	not with --refresh.

*--start-polling*
	Make a running astroid instance watch for changes in the mail directory and
	display a polling spinner. One must call --stop-polling at the end of the
	external polling (e.g. by offlineimap) *even if it fails*. --refresh can be
	used as an *alternative*, but not with --{start,stop}-polling.

*--stop-polling*
	Stop the polling initiated by --start-polling and update the user view with
	any changes detected since --start-polling was executed.

*-t, --test-config*
	Use test configuration file. Only makes sense from the source root.

*--disable-plugins*
	Disable all the plugins.

# CONFIGURATION

The configuration file location defaults to $XDG\_CONFIG\_HOME/astroid/config.
The preferred way to configure a new astroid setup is to use --new-config to
generate an initial configuration file, then edit it to match the user's needs.
You will find all the details on Astroid's online documentation:
https://github.com/astroidmail/astroid/wiki. If $NOTMUCH\_CONFIG is set it
takes precedence over the configured value for the notmuch database in the
configuration file.

You can customized the thread view by putting your own HTML and SCSS files in
$XDG\_CONFIG\_HOME/astroid/ui/, as explained on
https://github.com/astroidmail/astroid/wiki/Customizing-the-user-interface.

If $XDG\_CONFIG\_HOME is not set, it defaults to ~/.config.

# NOTE

Astroid is a *Mail User Agent*: all Astroid provides is a graphical interface
to your email (read, view, write, modify, organise). Thus, Astroid enables you
to launch actions that rely on the performance of other programs to actually
fetch, sync, index, search and send your email. It depends on Notmuch for
indexing and searching, but others actions can be handled many different
programs:

- Fetching and syncing your mail between your local computer and email service
provider can be handled by any program supporting the Maildir format:
	* *OfflineIMAP*
	* *mbsync*
	* *Gmailleer*
	* ...

- Sending mail through you email provider can be handler by any sendmail
compatible program:
	* *msmtp*
	* ...

- Editing can be handled by almost any editor:
	* *vim*
	* *emacs*
	* *gedit*
	* ...

While all these programs (or equivalent programs) need to be setup; once
they're configured, you shouldn't need to think about them beyond using
astroid. The configuration options may be more or less tedious and complicated,
depending on what email service provider you use.

# AUTHORS

Maintained by Gaute Hope <eg@gaute.vetsj.com>, who is assisted by other open
source contributors. For more information about astroid development, see
<https://github.com/astroidmail/astroid>.

# SEE ALSO

*notmuch*(1) *offlineimap*(1) *msmtp*(1)


History.txt
==

* Respect $NOTMUCH_CONFIG and ignore configuration option if set. Otherwise fall back to option from configuration file.

* We now have a man page! Generated either by scdoc or ronn.

== v0.13 / 2018-07-20

* Move to WebKit2

  We have now finally moved to WebKit2. This is a big change since the previous
  functionality had to be moved to a WebExtension running in a separate process.

  The IPC with the extension is implemented using Google ProtoBufs (>= 3.0)
  and Unix Domain Sockets.

  WebKit2 is actively maintained and consequently a long list of CVEs in the
  old WebKit no longer cause worries. Additionally, each part is shown in
  separate IFRAMEs so that essentially all received content is sandboxed (some
  metadata like addresses are escaped and shown in main page). This protects better
  against EFAIL type attacks (which I believe we were not vulnerable to anyway).

  Expect some instability while breaking in this new change!

  Important:  Make sure you update your plugins! If they try to use the old
              WebKit GIO bindings with the new WebKit objects astroid will freeze.

Part rendering:
* Show parts in IFRAME
* Toggle between parts without re-opening
* The default part shown (HTML or TEXT) is now configurable
* Optionally show HTML part if no other part

Images and EFAIL:
* Do not allow images in encrypted messages (#499)
* Show CID (attached) images in HTML parts (on C-i)
* Zoom messages with C-+ and C--

Syntax highlighting:
* Syntax highlighting through JS has been replaced by a syntax highlight
  plugin based on Pygments.

* Add filter_part plugin method, which can be used to modify the displayed HTML
  depending on either the text/plain or the text/html part.

Misc:
* Logging can be configured to go to syslog (or stdout, default). File logging
  is removed since we now have multiple processes and this would complicate
  writing to one file.

* The configuration file is no longer written back by default if missing values
  are found. This also means that values undefined by the user will be updated
  if the default is changed.

  Various checks for merging older versions of the configuration file to the
  current version has been removed. The parser is more resilient and the old
  versions are old.

== v0.12
* `mail.close_on_success` configures if page closes on successful send.
* `terminal.font_description` configures font of embedded terminal.
* `terminal.height` configures height of embedded terminal.
* `mail.format_flowed` is off by default.
* Make hooks undoable by providing a field for the inverse command.
* Do not force throw-keyid (follow gpg config).
* Allow GPG to be disabled with configuration option.

== v0.11.1 / 2018-02-20

* Automatically detect GMime version to build against.

== v0.11.0 / 2018-02-04

* Move to the CMake build system. Meson is deprecated.
* Allow HTML parts to be composed using Markdown.
* Use color to indicate whether a message is marked in thread-index.
* Always throw key-id when sending (using GMime 3)
* Bunch of minor stuff

== v0.10.2 / 2017-10-11

* Fix build issues on Mac OS.
* Gmime 2 used weird timezone for sent email.
* Add Meson build system as alternative.

== v0.10.1 / 2017-09-24

* Fix compilation errors against older versions of notmuch and gmime.

== v0.10 / 2017-09-21

* Support GMime 3 with backwards compatability with GMime 2. This orients us
  towards GMime 3, and support for GMime 2 will be dropped in the future.

* Attach messages by mids with 'A' in new emails.

* Colorful tags in tag editor and in searching

* Tags in the tag editor is now separated by spaces (comma still works)

* New configuration option 'signature_separate' determines if inline
  signatures will be separated from the message by a line containing '-- '.

* Configurable query associated with account can be used to select account as
  From: when replying or forwarding e-mail.

* Default clipboard target: CLIPBOARD since this is the only one accessible by keyboard
  in GTK.

* Select target for yank operations with '"'

* Track focus changes manually

== v0.9.1 / 2017-04-30

* Support old vte3.
* Support old _st versions on Arch.

== v0.9 / 2017-04-28

* Drop use of old deprecated non _st functions, use their new names. Will
  require 0.24 or newer.

* No longer support pre-lastmod notmuches.

* Catch missing content errors for messages, not fail.

* Wait and block R/O database opening, not fail.

* Defer changed_thread events on thread-index while still loading.

* Do not update tags in thread-view when not ready.

* configuration option thread_view.expand_flagged determines whether flagged
  messages are expanded by default.

* Better handle poll thread. Make sure to clean up connections here and in send_message.

* Polling can be cancelled with C-c in main window.

* Archive or close thread and then directly open next or next unread thread
  in originating thread-index.

== v0.8 / 2017-03-17

* New logo and icons by Elio Qoshi ( http://patreon.com/ura )

* Run hook for thread-view (Hugo Roy)

* Session-persistent folder history for attaching files and saving
  attachments. default is configurable.

* Remove duplicates when reply-all'ing.

* Handle mailto when astroid is already running.

* `scons --propagate-environment` uses environment from build environment.

* Show pane-focus with blue bar at top, swap focus with M-space. The
  keybinding is now located at 'pane.swap_focus'.

* Add optional drop-down terminal ('|') which has session-persistent
  working directory. Will silently be disabled if VTE is missing.

* Delay sending of message with configured time, in order to be able to
  cancel wrongly sent messages.

* New --start-polling, --stop-polling and --refresh LASTMOD options which
  makes external polling mechanisms work better.

* All notmuch tags are added to message CSS with 'nm-' prefix and some sanitation to allow thread-view customizing based on tag.

* Reply-To mailing list.


src/astroid.cc
# include <iostream>
# include <vector>
# include <memory>
# include <atomic>

# include <gtkmm.h>
# include <gtkmm/window.h>

/* program options */
# include <boost/program_options.hpp>
# include <boost/filesystem.hpp>

/* log */
# include <boost/log/core.hpp>
# include <boost/log/utility/setup/file.hpp>
# include <boost/log/utility/setup/console.hpp>
# include <boost/log/utility/setup/common_attributes.hpp>
# include <boost/log/sinks/text_file_backend.hpp>
# include <boost/log/sources/severity_logger.hpp>
# include <boost/log/sources/record_ostream.hpp>
# include <boost/log/utility/setup/filter_parser.hpp>
# include <boost/log/utility/setup/formatter_parser.hpp>
# include <boost/date_time/posix_time/posix_time_types.hpp>
# include <boost/log/expressions.hpp>
# include <boost/log/trivial.hpp>
# include <boost/log/support/date_time.hpp>
# include <boost/log/sinks/syslog_backend.hpp>

# include "astroid.hh"
# include "build_config.hh"
# include "db.hh"
# include "config.hh"
# include "account_manager.hh"
# include "actions/action_manager.hh"
# include "actions/action.hh"
# include "utils/date_utils.hh"
# include "utils/utils.hh"
# include "utils/resource.hh"

# ifndef DISABLE_PLUGINS
  # include "plugin/manager.hh"
# endif

# include "poll.hh"

/* UI */
# include "main_window.hh"
# include "modes/thread_index/thread_index.hh"
# include "modes/edit_message.hh"
# include "modes/saved_searches.hh"

/* gmime */
# include <gmime/gmime.h>
# include <utils/gmime/gmime-compat.h>

using namespace std;
using namespace boost::filesystem;

namespace logging   = boost::log;
namespace keywords  = boost::log::keywords;
namespace expr      = boost::log::expressions;

/* globally available static instance of the Astroid */
refptr<Astroid::Astroid> Astroid::astroid;
const char* const Astroid::Astroid::version = GIT_DESC;

namespace Astroid {
  // Initialization and creation {{{
  refptr<Astroid> Astroid::create () {
    return refptr<Astroid> (new Astroid ());
  }

  void Astroid::init_console_log () {
    /* log to console */
    logging::formatter format =
                  expr::stream
                      << "["
                      << expr::format_date_time< boost::posix_time::ptime >("TimeStamp", "%H:%M:%S")
                      << "] [" << expr::attr <boost::log::attributes::current_thread_id::value_type>("ThreadID")
                      << "] [M] [" << logging::trivial::severity
                      << "] " << expr::smessage
              ;

    logging::add_console_log ()->set_formatter (format);
  }

  void Astroid::init_sys_log () {
    typedef logging::sinks::synchronous_sink< logging::sinks::syslog_backend > sink_t;

    // Create a backend
    boost::shared_ptr< logging::sinks::syslog_backend > backend(new logging::sinks::syslog_backend(
          keywords::facility = logging::sinks::syslog::user,
          keywords::use_impl = logging::sinks::syslog::native,
          keywords::ident    = log_ident
          ));

    // Set the straightforward level translator for the "Severity" attribute of type int
    backend->set_severity_mapper(
        logging::sinks::syslog::direct_severity_mapping< int >("Severity"));

    // Wrap it into the frontend and register in the core.
    // The backend requires synchronization in the frontend.
    logging::core::get()->add_sink(boost::make_shared< sink_t >(backend));
  }

  Astroid::Astroid () :
    Gtk::Application("org.astroid",
        Gio::APPLICATION_HANDLES_OPEN | Gio::APPLICATION_HANDLES_COMMAND_LINE)
  {
    setlocale (LC_ALL, "");
    Glib::init ();

    /* user agent */
    user_agent = ustring::compose ("astroid/%1 (https://github.com/astroidmail/astroid)", Astroid::version);

    /* gmime settings */
    g_mime_init ();

    /* options */
    desc.add_options ()
      ( "help,h", "print this help message")
      ( "config,c", po::value<ustring>(), "config file, default: $XDG_CONFIG_HOME/astroid/config")
      ( "new-config,n",   "make new default config, then exit")
# ifdef DEBUG
      ( "test-config,t",  "use test config (same as used when tests are run), only makes sense from the source root")
# endif
      ( "mailto,m", po::value<ustring>(), "compose mail with mailto url or address")
      ( "no-auto-poll",   "do not poll automatically")
      ( "disable-log",    "disable logging")
      ( "log-stdout",     "log to stdout regardless of configuration")
      ( "log-level", po::value<ustring>(), "override configured logging level (trace, debug, info, warning, error, fatal)")
      ( "start-polling",  "indicate that external polling (external notmuch db R/W operations) starts")
      ( "stop-polling",   "indicate that external polling stops")
      ( "refresh", po::value<unsigned long>(), "refresh messages changed since lastmod")
# ifndef DISABLE_PLUGINS
      ( "disable-plugins", "disable plugins");
# else
      ;
# endif
  }
  // }}}

  int Astroid::run (int argc, char **argv) { // {{{
    LOG (info) << "welcome to astroid! - " << Astroid::version;
    register_application ();

    Resource::init (argv[0]);

    po::variables_map vm;

    bool show_help = false;

    try {
      po::store ( po::parse_command_line (argc, argv, desc), vm );
    } catch (po::unknown_option &ex) {
      LOG (error) << "unknown option" << endl;
      LOG (error) << ex.what() << endl;
      show_help = true;
    }

    show_help |= vm.count("help");
    bool test_config = vm.count("test-config");

    if (show_help) {

      cout << desc << endl;
      exit (0);

    }

    logging::add_common_attributes ();
    if (vm.count ("log-stdout")) {
      log_stdout = true;
      init_console_log ();
    }

    if (vm.count ("disable-log")) {
      logging::core::get()->set_logging_enabled (false);
      disable_log = true;
    }

    /* make new config {{{ */
    if (vm.count("new-config")) {
      if (test_config) {
        LOG (error) << "--new-config cannot be specified together with --test-config.";
        exit (1);
      }

      LOG (info) << "creating new config..";
      ustring cnf;

      Config ncnf (false, true);

      if (vm.count("config")) {
        cnf = vm["config"].as<ustring>().c_str();

        if (exists (path(cnf))) {
          LOG (error) << "the config file: " << cnf << " already exists.";
          exit (1);
        }

        ncnf.std_paths.config_file = path(cnf);

      } else {
        /* use default */
        if (exists(ncnf.std_paths.config_file)) {
          LOG (error) << "the config file: " << ncnf.std_paths.config_file.c_str() << " already exists.";
          exit (1);
        }
      }

      LOG (info) << "writing default config to: " << ncnf.std_paths.config_file.c_str();
      ncnf.load_config (true);
      ncnf.write_back_config ();

      exit (0);
    } // }}}

    if (!is_remote ()) {
      /* load config */
      if (vm.count("config")) {
        if (test_config) {
          LOG (error) << "--config cannot be specified together with --test-config.";
          exit (1);
        }

        LOG (info) << "astroid: loading config: " << vm["config"].as<ustring>().c_str();
        m_config = new Config (vm["config"].as<ustring>().c_str());
      } else {
        if (test_config) {
          m_config = new Config (true);
        } else {
          m_config = new Config ();
        }
      }

      /* setting up loggers */
      if (config ("astroid.log").get<bool>("stdout") && !vm.count ("log-stdout")) {
        init_console_log ();
        log_stdout = true;
        LOG (debug) << "log: stdout: yes";
      }

      if (config ("astroid.log").get<bool> ("syslog")) {
        init_sys_log ();
        log_syslog = true;
        LOG (debug) << "log: syslog: yes";
      }

      if (vm.count ("log-level")) {
        log_level = vm["log-level"].as<ustring> ();
      } else {
        log_level = config ("astroid.log").get<std::string> ("level");
      }

      /* Non existing llevel in map will be silently ignored */
      logging::core::get()->set_filter (logging::trivial::severity >= sevmap[log_level]);


      string charset;
      Glib::get_charset(charset);
      LOG (debug) << "utf8: " << Glib::get_charset () << ", " << charset;
      LOG (info) << "log: level: " << log_level;

      if (!Glib::get_charset()) {
        LOG (error) << "astroid needs an UTF-8 locale! this is probably not going to work.";
      }


      if (vm.count("start-polling") || vm.count("stop-polling") || vm.count ("refresh")) {
        LOG (error) << "--start-polling, --stop-polling or --refresh can only be specifed when there is already a running astroid instance.";
        exit (1);
      }

      _hint_level = config ("astroid.hints").get<int> ("level");

      /* set up classes */
      Date::init ();
      Utils::init ();

      /* Initialize Db and check if it has been set up */
      try {
        Db::init ();
        Db d; d.get_revision ();
      } catch (database_error &ex) {
        LOG (error) << "db: failed to open database, please check the manual if everything is set up correctly: " << ex.what ();

        Gtk::MessageDialog md ("Astroid: Failed to open database", false, Gtk::MESSAGE_ERROR);
        md.set_secondary_text (ustring::compose ("Please check the documentation if you have set up astroid and notmuch correctly.\n Error: <b>%1</b>", ex.what ()), true);

        md.run ();

        on_quit ();
        return 1;
      }

      Keybindings::init ();
      SavedSearches::init ();

      /* set up accounts */
      accounts = new AccountManager ();

# ifndef DISABLE_PLUGINS
      /* set up plugins */
      bool disable_plugins = vm.count ("disable-plugins");
      plugin_manager = new PluginManager (disable_plugins, in_test ());
      plugin_manager->astroid_extension = new PluginManager::AstroidExtension (this);
# endif

      /* set up global actions */
      actions = new ActionManager ();

      /* set up poller */
      bool no_auto_poll = false;
      if (vm.count ("no-auto-poll")) {
        LOG (info) << "astroid: automatic polling is off.";

        no_auto_poll = true;
      }
      poll = new Poll (!no_auto_poll);

      Gtk::Application::run (argc, argv);

      on_quit ();

    } else {
      Gtk::Application::run (argc, argv);

    }

    return 0;
  } // }}}

  const boost::property_tree::ptree& Astroid::config (const std::string& id) const {
    return m_config->config.get_child(id);
  }

  const boost::property_tree::ptree& Astroid::notmuch_config () const {
    return m_config->notmuch_config;
  }

  const StandardPaths& Astroid::standard_paths() const {
    return m_config->std_paths;
  }

  RuntimePaths& Astroid::runtime_paths() const {
    return m_config->run_paths;
  }

  bool Astroid::has_notmuch_config () {
    return m_config->has_notmuch_config;
  }

  void Astroid::main_test () { // {{{
    init_console_log ();

    m_config = new Config (true);

    /* set up static classes */
    Date::init ();
    Utils::init ();
    Db::init ();
    SavedSearches::init ();

    /* set up accounts */
    accounts = new AccountManager ();

# ifndef DISABLE_PLUGINS
    /* set up plugins */
    plugin_manager = new PluginManager (false, true);
    plugin_manager->astroid_extension = new PluginManager::AstroidExtension (this);
# endif

    /* set up contacts */
    //contacts = new Contacts ();

    /* set up global actions */
    actions = new ActionManager ();

    /* set up poller */
    poll = new Poll (false);
  } // }}}

  bool Astroid::in_test () {
    return m_config->test;
  }

  void Astroid::on_quit () {
    LOG (debug) << "astroid: quitting..";

    if (poll && poll->get_auto_poll ()) poll->toggle_auto_poll  ();
    if (poll) poll->close ();

    if (actions) actions->close ();
    SavedSearches::destruct ();

# ifndef DISABLE_PLUGINS
    if (plugin_manager && plugin_manager->astroid_extension) delete plugin_manager->astroid_extension;
    if (plugin_manager) delete plugin_manager;
# endif

    LOG (info) << "astroid: goodbye!";
  }

  Astroid::~Astroid () {
    if (accounts) delete accounts;

    if (m_config) delete m_config;

    if (poll) {
      poll->close ();
      delete poll;
    }

    if (actions) {
      actions->close ();
      delete actions;
    }
  }

  int Astroid::on_command_line (const refptr<Gio::ApplicationCommandLine> & cmd) {
    char ** argv;
    int     argc;
    bool new_window = true;

    argv = cmd->get_arguments (argc);

    if (get_windows().empty ()) {
      activate ();
      new_window = false;
    }

    /* handling command line arguments that may be passed from secondary
     * instances as well */
    if (argc > 0) {
      po::variables_map vm;

      try {
        po::store ( po::parse_command_line (argc, argv, desc), vm );
      } catch (po::unknown_option &ex) {
        LOG (error) << "unknown option" << endl;
        LOG (error) << ex.what() << endl;
        return 1;
      }

      if (vm.count("mailto")) {
        ustring mailtourl = vm["mailto"].as<ustring>();
        send_mailto (mailtourl);
        new_window = false;
      }

      if ((vm.count ("start-polling") ? 1:0) + (vm.count ("stop-polling") ? 1:0) + (vm.count("refresh") ? 1:0) > 1) {
          LOG (error) << "only one of --start-polling, --stop-polling and --refresh should be specified";
          return 1;
      }

      if (vm.count ("start-polling")) {

        poll->start_polling ();
        new_window = false;

      } else if (vm.count ("stop-polling")) {

        poll->stop_polling ();
        new_window = false;

      } else if (vm.count ("refresh")) {
        unsigned long last = vm["refresh"].as<unsigned long> ();
        poll->refresh ( last );
        new_window = false;
      }
    }

    if (new_window) activate ();

    return 0;
  }

  MainWindow * Astroid::open_new_window (bool open_defaults) {
    LOG (warn) << "astroid: starting a new window..";

    /* set up a new main window */

    /* start up default window with default buffers */
    MainWindow * mw = new MainWindow (); // is freed / destroyed by application

    if (open_defaults) {
      if (config ("saved_searches").get<bool>("show_on_startup")) {
        Mode * s = (Mode *) new SavedSearches (mw);
        s->invincible = true;
        mw->add_mode (s);
      }

      ptree qpt = config ("startup.queries");

      for (const auto &kv : qpt) {
        ustring name = kv.first;
        ustring query = kv.second.data();

        LOG (info) << "got query: " << name << ": " << query;
        Mode * ti = new ThreadIndex (mw, query, name);
        ti->invincible = true; // set startup queries to be invincible
        mw->add_mode (ti);
      }

      mw->set_active (0);
    }

    mw->signal_delete_event ().connect (
        sigc::bind (
          sigc::mem_fun(*this, &Astroid::on_window_close), mw));

    add_window (*mw);
    mw->show_all ();

    return mw;
  }

  bool Astroid::on_window_close (GdkEventAny *, MainWindow * mw) {
    // application automatically removes window

    // https://mail.gnome.org/archives/gtkmm-list/2004-March/msg00282.html
    delete mw;

    return false; // default Gtk handler destroys window
  }

  void Astroid::on_activate () {
    open_new_window ();
  }

  void Astroid::send_mailto (ustring url) {
    LOG (info) << "astroid: mailto: " << url;

    MainWindow * mw = (MainWindow*) get_windows ()[0];

    ustring scheme = Glib::uri_parse_scheme (url);
    if (scheme.length () > 0) {
      /* we got an mailto url */
      url = url.substr(scheme.length(), url.length () - scheme.length());

      ustring to, cc, bcc, subject, body;

      ustring::size_type pos = url.find ("?");
      /* ustring::size_type next; */
      if (pos == ustring::npos) pos = url.length ();
      to = url.substr (0, pos);

      /* TODO: need to finish the rest of the fields */
      mw->add_mode (new EditMessage (mw, to));

    } else {
      /* we probably just got the address on the cmd line */
      mw->add_mode (new EditMessage (mw, url));
    }
  }

  int Astroid::hint_level () {
    return _hint_level;
  }

}

